packages/connections/src/stores/connections-store.ts

```diff
@@ -0,0 +1,275 @@
+    },
+  };
+}
+
+type State = {
+  activeConnectionId?: string;
+  activeConnectionInfo: ConnectionInfo;
+  connectingStatusText: string;
+  connectionAttempt: ConnectionAttempt | null;
+  connectionErrorMessage: string | null;
+  connections: ConnectionInfo[];
+  isConnected: boolean;
+};
+
+export function defaultConnectionsState(): State {
+  return {
+    activeConnectionId: undefined,
+    activeConnectionInfo: createNewConnectionInfo(),
+    connectingStatusText: '',
+    connections: [],
+    connectionAttempt: null,
+    connectionErrorMessage: null,
+    isConnected: false,
+  };
+}
+
+type Action =
+  | {
+      type: 'attempt-connect';
+      connectionAttempt: ConnectionAttempt;
+      connectingStatusText: string;
+    }
+  | {
+      type: 'cancel-connection-attempt';
+    }
+  | {
+      type: 'connection-attempt-errored';
+      connectionErrorMessage: string;
+    }
+  | {
+      type: 'connection-attempt-succeeded';
+    }
+  | {
+      type: 'new-connection';
+      connectionInfo: ConnectionInfo;
+    }
+  | {
+      type: 'set-active-connection';
+      connectionId: string;
+      connectionInfo: ConnectionInfo;
+    }
+  | {
+      type: 'set-connections';
+      connections: ConnectionInfo[];
+    };
+
+export function connectionsReducer(state: State, action: Action): State {
+  switch (action.type) {
+    case 'attempt-connect':
+      return {
+        ...state,
+        connectionAttempt: action.connectionAttempt,
+        connectingStatusText: action.connectingStatusText,
+        connectionErrorMessage: null,
+      };
+    case 'cancel-connection-attempt':
+      return {
+        ...state,
+        connectionAttempt: null,
+        connectionErrorMessage: 'Connection attempt canceled.',
+      };
+    case 'connection-attempt-succeeded':
+      return {
+        ...state,
+        connectionAttempt: null,
+        isConnected: true,
+      };
+    case 'connection-attempt-errored':
+      return {
+        ...state,
+        connectionAttempt: null,
+        connectionErrorMessage: action.connectionErrorMessage,
+      };
+    case 'set-active-connection':
+      return {
+        ...state,
+        activeConnectionId: action.connectionId,
+        activeConnectionInfo: action.connectionInfo,
+      };
+    case 'new-connection':
+      return {
+        ...state,
+        activeConnectionId: action.connectionInfo.id,
+        activeConnectionInfo: action.connectionInfo,
+      };
+    case 'set-connections':
+      return {
+        ...state,
+        connections: action.connections,
+      };
+    default:
+      return state;
+  }
+}
+
+export function useConnections(
+  onConnected: (
+    connectionInfo: ConnectionInfo,
+    dataService: DataService
+  ) => Promise<void>
+): [
+  State,
+  {
+    onCancelConnectionAttempt(): void;
+    onConnect(connectionInfo: ConnectionInfo): Promise<void>;
+    updateActiveConnection(newConnectionId?: string | undefined): void;
```

Comment:
what's the right convention for these stuff? should this be:

```
connect
cancelConnectionAttempt
updateActiveConnection
```

Super random: `updateActiveConnection` at first made me think about something that would have really "changed" the connection info properties. Would something like `setActiveConnection` or `setActiveConnectionById` be clearer?

Reply:
FWIW as these are actions, it makes sense to name them `connect`, `cancelConnect`, `updateActive...`, etc